### ISSUE TITLE
Question: Has Flow-GRPO-Fast been tested on OCR tasks or GRPO-Guard? (Fast + GRPO-Guard seems unstable)

### ISSUE THREAD

**Issue by stephenjkhsieh** [yifan123/flow_grpo#192]
Hi, thanks for the great work!
I’m trying to apply **Flow-GRPO-Fast** + **GRPO-Guard** to an **OCR reward** task, but I found that the training becomes unstable compared to the original GRPO-Guard setup.

In the default configs, I didn’t find any OCR-related Flow-GRPO-Fast settings, so I’m wondering whether Fast is expected to work for OCR, or whether combining GRPO-Guard + Fast is not recommended.

Below is my observation:

----

**Experiment Setup**

- Hardware: 4 × H100
- Base model: SD3
- Task: OCR (same as `general_ocr_sd3_rationorm`)
- Batch setup:
  - `train_batch_size` = 8
  - `num_image_per_prompt` = 16
- GRPO-Guard baseline: uses the default config → training is stable
- Flow-GRPO-Fast attempt + GRPO-Guard:
Since there is no OCR Fast config, I follow `geneval_sd3_fast_nocfg`:
```python=
sde_window_size = 3
sde_window_range = (0, config.sample.num_steps // 2)
sde_type = "cps"
sample.guidance_scale = 4.5   # to keep comparability with the GRPO-Guard run
```
Everything else is identical to `general_ocr_sd3_rationorm`.

----
**Result**

The reward curve of **GRPO-Guard (green)** increases normally.
But **Fast** + **GRPO-Guard** (pink) quickly collapses:

<img width="1187" height="648" alt="Image" src="https://github.com/user-attachments/assets/b9692dc4-65d2-403c-b227-fbccd3e78f22" />

----
**My questions**

1. Have you tested Flow-GRPO-Fast on OCR tasks before?
  If yes, is it expected to behave very differently from GenEval / PickScore tasks?
3. Is Fast theoretically suitable for OCR rewards?
  Since OCR rewards rely heavily on fine-grained spatial/text details, maybe the reduced-step window in Fast harms the gradient quality?
4. Is the combination “GRPO-Guard + Fast” reasonable?
  Or is there any known incompatibility?

----
Thanks again for sharing the implementation and the such great paper.
Any guidance would be much appreciated!

**Comment by yifan123** [comment]
Please set sde_type = "sde". Currently, grpo-guard only supports flow-SDE and does not support flow-CPS.

**Comment by stephenjkhsieh** [comment]
Thanks a lot for the quick clarification!

After switching `sde_type` to `"sde"` as you suggested, the training became fully stable.
I reran the experiment and the results look great:

- Training speed improved by ~1.6×
- Eval reward improves faster than the original GRPO-Guard setup

Here are the updated curves (Fast + GRPO-Guard in orange, original GRPO-Guard in green):

<img width="1182" height="412" alt="Image" src="https://github.com/user-attachments/assets/7f9b128d-fdc0-4270-8929-99218e8913a6" />

Really appreciate your help — the fix worked perfectly!

**Comment by yifan123** [comment]
How many windows were used in your training? It seems that the "fast" performance isn't particularly noticeable. The OCR task is relatively simple, so the improvement might be more significant in other tasks.

**Comment by stephenjkhsieh** [comment]
I used the following settings:
```
config.sample.num_steps = 10
sde_window_size = 3
sde_window_range = (0, config.sample.num_steps // 2)
```
Just to confirm — when you mentioned “windows”, were you referring to the `sde_window_size` and `sde_window_range` here?

**Comment by MengHao666** [comment]
> Please set sde_type = "sde". Currently, grpo-guard only supports flow-SDE and does not support flow-CPS.

could u kindly add this notification in Readme.md? I combined "cps" with "sde" together in my recent experiments.

**Comment by happynear** [comment]
@MengHao666 
When using Flow-CPS, just remove all the denominators. 

In https://github.com/yifan123/flow_grpo/blob/main/scripts/train_sd3_GRPO_Guard.py#L918

```
                            ratio_mean_bias = ratio_mean_bias / (2 * (sqrt_dt.mean() * sigma_t) ** 2)
                            ratio = torch.exp((log_prob - sample["log_probs"][:, j] + ratio_mean_bias) * (sqrt_dt.mean() * sigma_t))
```
remove the first line and change the second line to
```
                            ratio = torch.exp(log_prob - sample["log_probs"][:, j] + ratio_mean_bias)
```

And there is no need to divide the denominator in policy loss:

https://github.com/yifan123/flow_grpo/blob/main/scripts/train_sd3_GRPO_Guard.py#L932

remove these two lines
```
                        if config.rationorm:
                            policy_loss = policy_loss / (sqrt_dt.mean()**2)
```

**Comment by MengHao666** [comment]
> https://github.com/yifan123/flow_grpo/blob/main/scripts/train_sd3_GRPO_Guard.py#L932

thanks, I will try and check the result.

**Comment by CIawevy** [comment]
> > https://github.com/yifan123/flow_grpo/blob/main/scripts/train_sd3_GRPO_Guard.py#L932
> 
> 谢谢，我会尝试并查看结果。

hi，结果如何呢

**Comment by MengHao666** [comment]
> > > https://github.com/yifan123/flow_grpo/blob/main/scripts/train_sd3_GRPO_Guard.py#L932
> > 
> > 
> > 谢谢，我会尝试并查看结果。
> 
> hi，结果如何呢

我没有见到非常大的差异，可能是我当前超参数设置  ratio超过范围被clip的情况不多。感觉flow-grpo受超参数影响非常大，非常敏感

**Comment by MengHao666** [comment]
> > > > https://github.com/yifan123/flow_grpo/blob/main/scripts/train_sd3_GRPO_Guard.py#L932
> > > 
> > > 
> > > 谢谢，我会尝试并查看结果。
> > 
> > 
> > hi，结果如何呢
> 
> 我没有见到非常大的差异，可能是我当前超参数设置 ratio超过范围被clip的情况不多



> [@MengHao666](https://github.com/MengHao666) When using Flow-CPS, just remove all the denominators.
> 
> In https://github.com/yifan123/flow_grpo/blob/main/scripts/train_sd3_GRPO_Guard.py#L918
> 
> ```
>                             ratio_mean_bias = ratio_mean_bias / (2 * (sqrt_dt.mean() * sigma_t) ** 2)
>                             ratio = torch.exp((log_prob - sample["log_probs"][:, j] + ratio_mean_bias) * (sqrt_dt.mean() * sigma_t))
> ```
> 
> remove the first line and change the second line to
> 
> ```
>                             ratio = torch.exp(log_prob - sample["log_probs"][:, j] + ratio_mean_bias)
> ```
> 
> And there is no need to divide the denominator in policy loss:
> 
> https://github.com/yifan123/flow_grpo/blob/main/scripts/train_sd3_GRPO_Guard.py#L932
> 
> remove these two lines
> 
> ```
>                         if config.rationorm:
>                             policy_loss = policy_loss / (sqrt_dt.mean()**2)
> ```

Have u check the result when using both grpo-guard and cps? Will they have innner confilicts?
